comments, symbols, metadata

guillermooo opened the pull request.

FichteFoll commented:
First, I think "comments" and "symbols" should not be a part of the _reference_ but of the user guide part (which is currently "extensibility" and before).

FichteFoll commented on source/reference/comments.rst:
```diff
@@ -0,0 +1,175 @@
+and uses the Property List format.
+The file name can be arbitrary.
+
+**Note:** Comment metadata is declared simply as ``shellVariables``.
```
This reads weird. How about:

```
Like general metadata,
comment metadata is stored in :file:`.tmPreferences` files
and uses the Property List format.
The file name can be arbitrary.

Because comment metadata is commonly required by packages
it's included in a separate topic
where it's easier to find.

**Note:** Comment metadata is declared inside a ``shellVariables`` dictionary.
```

guillermooo replied:
done

FichteFoll commented on source/reference/comments.rst:
```diff
@@ -0,0 +1,175 @@
+               <key>name</key>
+               <string>TM_COMMENT_START</string>
+               <key>value</key>
+               <string>/// </string>
```
How about a real world example here (this is only an excerpt)? I recommend Python's `<string>#</string>` as it's a common language with no block comments, which will be introduced later.

guillermooo replied:
done

FichteFoll commented on source/reference/comments.rst:
```diff
@@ -0,0 +1,175 @@
+   </dict>
+   </plist>
+
+These are all valid elements
```
In the following we should only list the valid top-level elements. `shellVariables` is not a top-level element and should be declared in the `settings` section.

`TM_*` should then be in the shellVariables section, or to make things easier in its own section with a different heading like "The following `shellVariables` settings can be specified for comment metadata".

guillermooo replied:
done

FichteFoll commented on source/reference/comments.rst:
```diff
@@ -0,0 +1,175 @@
+   <key>value</key>
+   <string>no</string>
+
+
```
There should be a complete example with a line and a block comment syntax at the end (so here). JavaScript is a nice candidate.

guillermooo replied:
todo

FichteFoll commented on source/reference/comments.rst:
```diff
@@ -0,0 +1,175 @@
+you can use default Sublime Text key bindings
+to comment/uncomment lines of code.
+
+To toggle a line comment, press :kbd:`Ctrl+/`
```
These two lines would be in a block text. I suggest an unordered list.

guillermooo replied:
done

FichteFoll commented on source/reference/metadata.rst:
```diff
@@ -0,0 +1,128 @@
+File Format
+===========
+
+General metadata files have the :file:`.tmPreferences` extension and use the
```
Just `Metadata files`. Preferrably everywhere.

guillermooo replied:
Agreed.

FichteFoll commented on source/reference/metadata.rst:
```diff
@@ -0,0 +1,128 @@
+
+
+``shellVariables``
+   Required. Contains other elements.
```
It's not clear enough that this should be a child of `settings`.

FichteFoll commented on source/reference/metadata.rst:
```diff
@@ -0,0 +1,128 @@
+   <array>
+   </array>
+
+``SOME_VARIABLE_NAME``
```
As with comments, it's not clear enough that these should be childs of `shellVariables` either. Or well, you write it in the definition but imo this needs better structuring.

FichteFoll commented on source/reference/metadata.rst:
```diff
@@ -0,0 +1,128 @@
+Indentation Options
+===================
+
+``increaseIndentPattern``
```
I suspect this is a WIP, but I think the `shellVariables`, since they are pretty specific, should go below this.

FichteFoll commented on source/reference/metadata.rst:
```diff
@@ -0,0 +1,128 @@
+Overview
+========
+
+General metadata files can be used to control several aspects of Sublime Text.
```
We need a general explanation of what metadata actually is, how it is assigned and what it can be used for.

Example:

```
Metadata are parameters
that can be assigned to certain text sections
using `scope selectors`_.

These parameters can for example be used for:

- specifying the current comment syntax,
  even with embedded source code,
  so that you can toggle comments in any syntax,
- defining indentation rules for auto-indenting source code,
- selecting symbols that ST will allow you to browse to quickly (x-ref goto anything).

.. Link to the separate comment and symbol sections from here

Furthermore, snippets have access to metadata specified in the ``shellVariables`` setting
which allows you to create a snippet
that has different contents
depending on where it's used.
```

Another thing I noticed: We should mention that syntax `includes` that cross-refernce other syntax files do NOT add its base scope name to the scope. So, for example with HTML including PHP,

``` yaml
- name: source.php
  include: source.php
```

Otherwise the concept of metadata wouldn't work. This doesn't belong in this topic but just mentioning.

FichteFoll commented on source/reference/symbols.rst:
```diff
@@ -0,0 +1,226 @@
+
+Symbol definition files have the :file:`.tmPreferences` extension
+and use the Property List format.
+Their file can be arbitrary.
```
file name

FichteFoll commented on source/reference/symbols.rst:
```diff
@@ -0,0 +1,226 @@
+
+.. TODO: Is that so (below)?
+
+Note that the ``showInSymbolList`` setting is required.
```
Instead of wriitng this into a separate note, I think we can embed this information into the paragraph above.

FichteFoll commented on source/reference/symbols.rst:
```diff
@@ -0,0 +1,226 @@
+before they are displayed to the user.
+Text transformations are defined
+as regular expressions.
+The regular expression engine used is `Oniguruma`_.
```
~~The possibility of specifying multiple semicolon-separated regex replacements is not mentioned here.~~

But later.

FichteFoll commented on source/reference/symbols.rst:
```diff
@@ -0,0 +1,226 @@
+``name``
+   Name of the symbol definition.
+
+.. XXX: Pretty useless, I believe.
```
The `name` key is probably useless for all of these since they are a relict from TM I suppose. If it's not required by ST (which I think not), we should mark it as optional.

FichteFoll commented on source/reference/symbols.rst:
```diff
@@ -0,0 +1,226 @@
+Valide elements inside ``settings``:
+
+``showInSymbolList``
+   ``0`` or ``1``. If ``0``, the corresponding symbols
```
Should mention that this is an integer type.

FichteFoll commented on source/reference/symbols.rst:
```diff
@@ -0,0 +1,226 @@
+      </string>
+   </dict>
+
+.. TODO: Are there more settings/options?
```
`symbolIndexTransformation`

FichteFoll commented on source/reference/symbols.rst:
```diff
@@ -0,0 +1,226 @@
+Navigating Symbols
+==================
+
+Once symbols are defined,
```
x-ref to goto anything

FichteFoll commented:
Just noticed that you didn't mention the `uuid`. It's not used by ST but we should mention it at least.

guillermooo commented:
> First, I think "comments" and "symbols" should not be a part of the reference but of the user guide part (which is currently "extensibility" and before).

Not sure about that. The actual technical reference part should remain under reference. We can have a short intro topic too in a different section, just as we do for other topìcs.

guillermooo commented:
@FichteFoll PTAL. All relevant info about metadata should be included now.

guillermooo committed:
add metadata topics

guillermooo committed:
minor fixes

guillermooo committed:
edit metadata topic

FichteFoll commented:
For the record: I made extensive passes on metadata and comments docs, symbols still to be done. Will see if I get to commit/push this tomorrow, Friday otherwise.

I also added a few items to our trello checklists for other topics.

guillermooo commented:
> For the record: I made extensive passes on metadata and comments docs, symbols still to be done. Will see if I get to commit/push this tomorrow, Friday otherwise.

Same: I still have a few edits to make. Will try to push them tonight or tomorrow.

guillermooo committed:
edit

FichteFoll committed:
edit metadata topics

more x-refs, code highlighting for xml, properly indented code blocks,
various other adjustments (including rephrasing some sections)

FichteFoll commented:
@guillermooo PTAL. I didn't do a final review on the rendered thing yet, but you can check it already.

Also, sidebar stuff is still missing. I'd need to do some digging in the forums there.

FichteFoll committed:
Also add a "seealso" block to the symbols doc

FichteFoll committed:
Address feedback from 7bf1ec38 (#102)

FichteFoll committed:
"document" -> "page"

guillermooo commented on source/reference/metadata.rst:
```diff
@@ -0,0 +1,315 @@
+- specifying the current comment markers,
+  even within embedded source code,
+  so that you can toggle comments in any syntax,
+- defining rules for auto-indentation,
```
http://www.chicagomanualofstyle.org/16/images/ch07_tab01.pdf

Can't get a hold of a CMS copy, but it seems auto- should not have a hyphen.

FichteFoll committed:
Very minor adjustments

From ccb7f2b feedback.
